### PR TITLE
[FLINK-25486][Runtime/Coordination] Fix the bug that flink will lost state when zookeeper leader changes

### PR DESCRIPTION
**Pull request by Myracle**
## What is the purpose of the change

*When the config high-availability.zookeeper.client.tolerate-suspended-connections is default false, the appMaster will failover once zk leader changes. In this case, the old appMaster will clean up all the HA data and the new appMaster will not recover from the latest checkpoint. We will fix it.*

## Brief change log

  - *When cluster is shutting down with ApplicationStatus.UNKNOWN, cleanupHaData will be set false to retain HA data.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates that HA data will not be cleanup after the cluster finished with ApplicationStatus.UNKNOWN.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by dmvk** [comment]
CI fails with https://issues.apache.org/jira/browse/FLINK-25307

**Comment by dmvk** [comment]
@flinkbot run azure

**Review by dmvk — CHANGES_REQUESTED** [review]
Thanks for the PR @Myracle! I think we need to take a slightly different approach here. Lost of the `JobMaster` leadership (see `JobMasterServiceLeadershipRunner`) shouldn't lead to the process termination. Instead the `JobMasterServiceLeadershipRunner` should be able to re-participate in the new election and potentially get re-elected.

I guess the fix could be as simple as not completing the shutdownFuture in `MiniDispatcher#jobReachedTerminalState` for non-globally terminal states (already suggested by @tillrohrmann in the JIRA).

WDYT?

**Comment by Myracle** [comment]
@dmvk Thanks for the review and suggestions. I agree with you that the the process should not exit in this case. I have modified the code. Could you review again?

**Comment by Myracle** [comment]
@dmvk Could you review again? Thank you.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/MiniDispatcherTest.java [hunk 1]
```diff
@@ -177,6 +177,34 @@ public void testTerminationAfterJobCompletion() throws Exception {
         }
     }
 
+    /**
+     * Tests that in detached mode, the {@link MiniDispatcher} will not complete the future that
+     * signals job termination if the JobStatus is not globally terminal state.
+     */
+    @Test
+    public void testNotTerminationWithoutGloballyTerminalState() throws Exception {
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/MiniDispatcherTest.java:185]: the test passes without the fix as well

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java [hunk 2]
```diff
@@ -124,16 +125,23 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
...
             // retrieval
             log.info(
                     "Shutting down cluster with state {}, jobCancelled: {}, executionMode: {}",
-                    archivedExecutionGraph.getState(),
+                    jobStatus,
+                    jobCancelled,
+                    executionMode);
+            shutDownFuture.complete(ApplicationStatus.fromJobStatus(jobStatus));
+        } else {
+            log.warn(
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java:140]: Why do we need to print a warning here?

Always try to think about it from the user perspective. If users sees this log message, is it relevant to what's going on? Is it something he / she should investigate?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java [hunk 3]
```diff
@@ -124,16 +125,23 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
                 executionGraphInfo.getArchivedExecutionGraph();
         final CleanupJobState cleanupHAState = super.jobReachedTerminalState(executionGraphInfo);
 
-        if (jobCancelled || executionMode == ClusterEntrypoint.ExecutionMode.DETACHED) {
+        JobStatus jobStatus = archivedExecutionGraph.getState();
+        if ((jobStatus != null && jobStatus.isGloballyTerminalState())
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java:129]: Can the job status be null here? My intuition would be that for the terminal state this should never be the case and would actually signal an underlying issue. Maybe adding a safeguard (`Objects.requireNonNull`) with a reasonable message could a better fit?

**Review by dmvk — CHANGES_REQUESTED** [review]
Thanks for updating the PR, this is headed in a good direction, I've added few more comments.

Even though I think this fixes the underlying issue, I would feel more comfortable if we could create a simple integration test for the actual scenario that has been reported. This would ensure that we don't introduce the same problem in the future by accident and that we've really fixed the problem.

Basically we'd test the following:
- Setup a `MiniCluster` with the `JobDispatcherFactory` and the corresponding job
- Wait for the checkpoint
- Revoke & Grant leadership
- Assert that we have recovered from checkpoint
- The test should fail without a fix (this currently doesn't hold for the newly introduced unit test, so this made me think that we need to put more effort here)

We already have a pretty similar test for the application mode, which you can use to as an inspiration (`ApplicationDispatcherBootstrapITCase`).

WDYT?

**Comment by tillrohrmann** [comment]
What's the state of this PR? Can we resolve the open comments to merge it soon?

**Comment by Myracle** [comment]
@tillrohrmann @dmvk I have just updated the code. Hope to merge it soon. Thank you.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/MiniDispatcherTest.java (hunk 1 above)
Comment (Myracle) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/MiniDispatcherTest.java:184]: I have modified the code to test the shutDownFuture not finished.

**Comment by Myracle** [comment]
> 

It is a good suggestion and I have followed it.

**Comment by Myracle** [comment]
@flinkbot run azure

**Comment by Myracle** [comment]
@flinkbot run azure

**Review by dmvk — COMMENTED** [review]
Thanks for the update @Myracle, I really like it! <3 I've added few more suggestions for the test cases for speeding them up + aligning on the community code style (junit 5 for the new tests).

I think we once we address these the PR should be good to go! Once more, thanks for the contribution, this is a really great improvement.

**Comment by Myracle** [comment]
@dmvk Thanks for the valuable suggestions which make the code clean. I have modified the code. Could you please review again?

**Review by dmvk — APPROVED** [review]
LGTM 👍 Thanks for the update

**Comment by dmvk** [comment]
:( compile step is failing, `AT_LEAST_ONE_CHECKPOINT_COMPLETED` needs to be renamed (-> camel case). Sorry for confusing you.

**Comment by Myracle** [comment]
@dmvk Fixed and the CI passed.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating the PR @Myracle. Merging it now.

**Pull request closed without merging**